firebase-database/src/main/java/com/google/firebase/database/core/Repo.java

```diff
@@ -548,15 +549,42 @@ public void run() {
                           source.setException(Objects.requireNonNull(task.getException()));
                         }
                       } else {
+                        /*
+                         We need to replicate the behavior that occurs when running `once()`. In other words,
+                         we need to create a new eventRegistration, register it with a view and then
+                         overwrite the data at that location, and then remove the view.
+                        */
                         Node serverNode = NodeUtilities.NodeFromJSON(task.getResult());
-                        postEvents(
-                            serverSyncTree.applyServerOverwrite(query.getPath(), serverNode));
+                        QuerySpec spec = query.getSpec();
+                        // EventRegistrations require a listener to be attached, so a dummy
+                        // ValueEventListener was created.
+                        ValueEventListener listener =
+                            new ValueEventListener() {
+                              @Override
+                              public void onDataChange(@NonNull DataSnapshot snapshot) {
+                                // noOp
+                              }
+
+                              @Override
+                              public void onCancelled(@NonNull DatabaseError error) {
+                                // noOp
+                              }
+                            };
+                        ValueEventRegistration eventRegistration =
+                            new ValueEventRegistration(repo, listener, spec);
+                        serverSyncTree.addEventRegistration(eventRegistration, true);
```

Comment:
nit: `serverSyncTree.addEventRegistration(eventRegistration, /*skipListenerSetup=*/true);`

Intellij does this for you but I think it's still useful when browsing on github.